p4-16/psa/examples/psa-example-register2.p4

```diff
@@ -0,0 +1,158 @@
+/*
+Copyright 2017 Cisco Systems, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include <core.p4>
+#include "../psa.p4"
+
+
+typedef bit<48>  EthernetAddress;
+
+header ethernet_t {
+    EthernetAddress dstAddr;
+    EthernetAddress srcAddr;
+    bit<16>         etherType;
+}
+
+header ipv4_t {
+    bit<4>  version;
+    bit<4>  ihl;
+    bit<8>  diffserv;
+    bit<16> totalLen;
+    bit<16> identification;
+    bit<3>  flags;
+    bit<13> fragOffset;
+    bit<8>  ttl;
+    bit<8>  protocol;
+    bit<16> hdrChecksum;
+    bit<32> srcAddr;
+    bit<32> dstAddr;
+}
+
+struct fwd_metadata_t {
+}
+
+struct metadata {
+    fwd_metadata_t fwd_metadata;
+}
+
+struct headers {
+    ethernet_t       ethernet;
+    ipv4_t           ipv4;
+}
+
+// BEGIN:Register_Example2_Part1
+const PortId_t NUM_PORTS = 512;
+
+// It would be more convenient to use a struct type to represent the
+// state of a combined packet and byte count, and many other compound
+// values one might wish to store in a Register instance.  However,
+// the latest p4test as of 2017-Aug-13 does not allow a struct type to
+// be returned from a method call like Register.read().
```

Comment:
I'm not sure I understand your comment.

A Register instantiation can specify whatever number of bits it wants per entry in the Register.  Whether that looks like a bit<40> or a struct with a bit<30> and a bit<10> in the P4_16 source code shouldn't matter to the hardware.  It is a nice convenience for a P4 program writer to have the second option available, though, yes?